orttraining/orttraining/core/session/training_session.cc

```diff
@@ -141,6 +153,40 @@ void TrainingSession::FilterUnusedWeights(const std::unordered_set<std::string>&
 
 const std::string TrainingSession::training_mode_string_ = "training_mode";
 
+// Create NCCL's communication plan. In runtime, we will provide details such
+// as pointer to sent/recieved data and the size of the data in byte. See how
+// Send and Recv call SubmitSendAndWait and SubmitRecvAndWait, respectively.
+void TrainingSession::LaunchNcclService(const int pipeline_stage_id) {
+  auto& nccl_service = cuda::NcclService::GetInstance();
```

Comment:
Think we need some separation here between the general code in training_session, and the NCCL-specific code.  We'd like to be able to run on CPU-only MPI (although I appreciate there are gaps in the set of operators that exist there right now), but we will also want to run on accelerators without NCCL.

Reply:
The content of this function is wrapped inside
```
#if defined(USE_NCCL) && defined(USE_NCCL_P2P)
  ...
#else
  ORT_UNUSED_PARAMETER(pipeline_stage_id);
#endif
```

---
In reply to: [527647827](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r527647827) [](ancestors = 527647827)